lib/RefBackend/RefBackend.cpp

```diff
@@ -95,28 +95,45 @@ static LogicalResult mungeFunction(
   SmallVector<Operation *> toErase;
   bool hadError = false;
   func.walk([&](ReturnOp op) {
-    auto memRefType = op.getOperandTypes()[0].dyn_cast<MemRefType>();
-    if (!memRefType) {
+    auto returnType = op.getOperandTypes()[0];
+    if (returnType.isa<MemRefType>()) {
+      auto elemType = returnType.dyn_cast<MemRefType>().getElementType();
+      auto unRankedMemrefType = UnrankedMemRefType::get(elemType, 0);
+      auto it = consumeFuncReturnFuncs.find(unRankedMemrefType);
+      if (op.getNumOperands() != 1 || it == consumeFuncReturnFuncs.end()) {
+        hadError = true;
+        op.emitError("must have one return value");
+        return;
+      }
+      b.setInsertionPoint(op);
+      auto cast = b.create<memref::CastOp>(
+          op.getLoc(), op.getOperand(0),
+          getAbiTypeForMemRef(op.getOperandTypes()[0]));
+      b.create<mlir::CallOp>(op.getLoc(),
+                             consumeFuncReturnFuncs[unRankedMemrefType],
+                             cast.getResult());
+      b.create<mlir::ReturnOp>(op.getLoc());
+      toErase.push_back(op);
```

Comment:
nit: the `CallOp`, `ReturnOp`, and `toErase` calls are pretty similar for all `if` branches. Just change `consumeFuncReturnFuncs[...]` with `it->second`, and have each branch define the third argument for `CallOp` and you should be able to factor this out.